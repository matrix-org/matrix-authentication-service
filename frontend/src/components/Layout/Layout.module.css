/* Copyright 2023 The Matrix.org Foundation C.I.C.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.layout-container {
  box-sizing: border-box;
  display: flex;
  flex-direction: column;

  max-width: calc(420px + var(--cpd-space-5x) * 2);

  /* Fallback for browsers that do not support 100svh */
  min-height: 100vh;
  min-height: 100svh;

  margin: 0 auto;
  padding-inline: var(--cpd-space-5x);
  padding-block: var(--cpd-space-12x);
  gap: var(--cpd-space-8x);

  &.wide {
    max-width: calc(520px + var(--cpd-space-5x) * 2);
  }
}

@media screen and (min-width: 768px) {
  .layout-container {
    padding-block-start: var(--cpd-space-20x);
    padding-block-end: var(--cpd-space-10x);
  }
}
