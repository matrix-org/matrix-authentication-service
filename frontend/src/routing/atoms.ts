// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { atom } from "jotai";
import { atomWithLocation } from "jotai-location";

import { Location, pathToRoute, Route, routeToPath } from "./routes";

export const appConfigAtom = atom<AppConfig>(
  typeof window !== "undefined" ? window.APP_CONFIG : { root: "/" },
);

const locationToRoute = (root: string, location: Location): Route => {
  if (!location.pathname || !location.pathname.startsWith(root)) {
    throw new Error(`Invalid location ${location.pathname}`);
  }

  const path = location.pathname.slice(root.length);
  return pathToRoute(path);
};

export const locationAtom = atomWithLocation();

export const routeAtom = atom(
  (get) => {
    const location = get(locationAtom);
    const config = get(appConfigAtom);
    return locationToRoute(config.root, location);
  },
  (get, set, value: Route, searchParams?: URLSearchParams) => {
    const appConfig = get(appConfigAtom);
    set(locationAtom, {
      pathname: appConfig.root + routeToPath(value),
      searchParams,
    });
  },
);
