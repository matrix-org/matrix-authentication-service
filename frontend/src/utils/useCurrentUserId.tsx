// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { useQuery } from "urql";

import { graphql } from "../gql";

const CURRENT_VIEWER_QUERY = graphql(/* GraphQL */ `
  query CurrentViewerQuery {
    viewer {
      __typename
      ... on User {
        id
      }
    }
  }
`);

export const useCurrentUserId = (): string | null => {
  const [result] = useQuery({ query: CURRENT_VIEWER_QUERY });
  if (result.error) throw result.error;
  if (!result.data) throw new Error(); // Suspense mode is enabled
  return result.data.viewer.__typename === "User"
    ? result.data.viewer.id
    : null;
};
