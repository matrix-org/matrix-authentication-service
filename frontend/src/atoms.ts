// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { AnyVariables, CombinedError, OperationContext } from "@urql/core";
import { atom, WritableAtom } from "jotai";
import { useHydrateAtoms } from "jotai/utils";
import { AtomWithQuery, atomWithQuery, clientAtom } from "jotai-urql";
import type { ReactElement } from "react";

import { graphql } from "./gql";
import { client } from "./graphql";
import { err, ok, Result } from "./result";

export type GqlResult<T> = Result<T, CombinedError>;
export type GqlAtom<T> = WritableAtom<
  Promise<GqlResult<T>>,
  [context?: Partial<OperationContext>],
  void
>;

/**
 * Map the result of a query atom to a new value, making it a GqlResult
 *
 * @param queryAtom: An atom got from atomWithQuery
 * @param mapper: A function that takes the data from the query and returns a new value
 */
export const mapQueryAtom = <Data, Variables extends AnyVariables, NewData>(
  queryAtom: AtomWithQuery<Data, Variables>,
  mapper: (data: Data) => NewData
): GqlAtom<NewData> => {
  return atom(
    async (get): Promise<GqlResult<NewData>> => {
      const result = await get(queryAtom);
      if (result.error) {
        return err(result.error);
      }

      if (result.data === undefined) {
        throw new Error("Query result is undefined");
      }

      return ok(mapper(result.data));
    },

    (_get, set, context) => {
      set(queryAtom, context);
    }
  );
};

export const HydrateAtoms: React.FC<{ children: ReactElement }> = ({
  children,
}) => {
  useHydrateAtoms([[clientAtom, client]]);
  return children;
};

const CURRENT_VIEWER_QUERY = graphql(/* GraphQL */ `
  query CurrentViewerQuery {
    viewer {
      __typename
      ... on User {
        id
      }

      ... on Anonymous {
        id
      }
    }
  }
`);

const currentViewerAtom = atomWithQuery({ query: CURRENT_VIEWER_QUERY });

export const currentUserIdAtom: GqlAtom<string | null> = mapQueryAtom(
  currentViewerAtom,
  (data) => {
    if (data.viewer.__typename === "User") {
      return data.viewer.id;
    }
    return null;
  }
);

const CURRENT_VIEWER_SESSION_QUERY = graphql(/* GraphQL */ `
  query CurrentViewerSessionQuery {
    viewerSession {
      __typename
      ... on BrowserSession {
        id
      }

      ... on Anonymous {
        id
      }
    }
  }
`);

const currentViewerSessionAtom = atomWithQuery({
  query: CURRENT_VIEWER_SESSION_QUERY,
});

export const currentBrowserSessionIdAtom: GqlAtom<string | null> = mapQueryAtom(
  currentViewerSessionAtom,
  (data) => {
    if (data.viewerSession.__typename === "BrowserSession") {
      return data.viewerSession.id;
    }
    return null;
  }
);
