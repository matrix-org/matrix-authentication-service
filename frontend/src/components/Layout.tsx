// Copyright 2022 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Link } from "@vector-im/compound-web";
import { useAtomValue } from "jotai";

import { currentUserIdAtom } from "../atoms";
import { isErr, unwrapErr, unwrapOk } from "../result";

import GraphQLError from "./GraphQLError";
import styles from "./Layout.module.css";
import NavBar from "./NavBar";
import NavItem from "./NavItem";
import NotLoggedIn from "./NotLoggedIn";
import UserGreeting from "./UserGreeting";

const Layout: React.FC<{ children?: React.ReactNode }> = ({ children }) => {
  const result = useAtomValue(currentUserIdAtom);
  if (isErr(result)) return <GraphQLError error={unwrapErr(result)} />;

  const userId = unwrapOk(result);
  if (userId === null)
    return (
      <div className={styles.container}>
        <NotLoggedIn />
      </div>
    );

  return (
    <div className={styles.container}>
      <UserGreeting userId={userId} />

      <NavBar>
        <NavItem route={{ type: "home" }}>Home</NavItem>
        <NavItem route={{ type: "email-list" }}>Emails</NavItem>
      </NavBar>

      <main>{children}</main>

      <footer className={styles.footer}>
        <nav className={styles.footerLinks}>
          <ul>
            <Link href="https://matrix.org/legal/copyright-notice">Info</Link>
            <Link href="https://matrix.org/legal/privacy-notice">Privacy</Link>
            <Link href="https://matrix.org/legal/terms-and-conditions">
              Terms & Conditions
            </Link>
          </ul>
        </nav>
      </footer>
    </div>
  );
};

export default Layout;
