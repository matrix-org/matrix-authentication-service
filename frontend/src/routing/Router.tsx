// Copyright 2022, 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { useAtom, useAtomValue } from "jotai";
import { useEffect } from "react";
import { useTranslation } from "react-i18next";

import LoadingSpinner from "../components/LoadingSpinner";
import BrowserSession from "../pages/BrowserSession";
import BrowserSessionList from "../pages/BrowserSessionList";
import OAuth2Client from "../pages/OAuth2Client";
import Profile from "../pages/Profile";
import SessionDetail from "../pages/SessionDetail";
import SessionsOverview from "../pages/SessionsOverview";
import VerifyEmail from "../pages/VerifyEmail";

import { getRouteActionRedirection } from "./actions";
import { locationAtom, routeAtom } from "./atoms";
import type { Route } from "./routes";

/**
 * Check for actions in URL query params requiring a redirect
 * Get route from path
 * @returns Route
 */
const useRouteWithRedirect = (): [Route, boolean] => {
  const location = useAtomValue(locationAtom);
  const redirect = getRouteActionRedirection(location);

  const [route, setRoute] = useAtom(routeAtom);

  useEffect(() => {
    if (redirect) {
      setRoute(redirect.route, redirect.searchParams);
    }
  }, [redirect, setRoute]);

  const redirecting = !!redirect;
  return [route, redirecting];
};

// A type-safe way to ensure we've handled all routes
const unknownRoute = (route: never): never => {
  throw new Error(`Invalid route: ${JSON.stringify(route)}`);
};

const Router: React.FC<{ userId: string }> = ({ userId }) => {
  const [route, redirecting] = useRouteWithRedirect();
  const { t } = useTranslation();

  if (redirecting) {
    return <LoadingSpinner />;
  }

  switch (route.type) {
    case "profile":
      return <Profile userId={userId} />;
    case "sessions-overview":
      return <SessionsOverview />;
    case "session":
      return <SessionDetail userId={userId} deviceId={route.id} />;
    case "browser-session-list":
      return <BrowserSessionList userId={userId} />;
    case "client":
      return <OAuth2Client id={route.id} />;
    case "browser-session":
      return <BrowserSession id={route.id} />;
    case "verify-email":
      return <VerifyEmail id={route.id} />;
    case "unknown":
      return (
        <>
          {t("frontend.unknown_route", {
            route: JSON.stringify(route.segments),
          })}
        </>
      );
    default:
      unknownRoute(route);
  }
};

export default Router;
