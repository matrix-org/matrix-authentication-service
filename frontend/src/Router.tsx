// Copyright 2022, 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { atom, useAtomValue, useSetAtom } from "jotai";
import { atomWithLocation } from "jotai-location";
import { lazy, Suspense, useTransition } from "react";

import styles from "./Router.module.css";
import Layout from "./components/Layout";
import LoadingSpinner from "./components/LoadingSpinner";

type Location = {
  pathname?: string;
  searchParams?: URLSearchParams;
};

type ProfileRoute = { type: "profile" };
type SessionOverviewRoute = { type: "sessions-overview" };
type SessionDetailRoute = { type: "session"; id: string };
type OAuth2ClientRoute = { type: "client"; id: string };
type OAuth2SessionList = { type: "oauth2-session-list" };
type BrowserSessionRoute = { type: "browser-session"; id: string };
type BrowserSessionListRoute = { type: "browser-session-list" };
type CompatSessionListRoute = { type: "compat-session-list" };
type VerifyEmailRoute = { type: "verify-email"; id: string };
type UnknownRoute = { type: "unknown"; segments: string[] };

export type Route =
  | SessionOverviewRoute
  | SessionDetailRoute
  | ProfileRoute
  | OAuth2ClientRoute
  | OAuth2SessionList
  | BrowserSessionRoute
  | BrowserSessionListRoute
  | CompatSessionListRoute
  | VerifyEmailRoute
  | UnknownRoute;

const routeToSegments = (route: Route): string[] => {
  switch (route.type) {
    case "profile":
      return [];
    case "sessions-overview":
      return ["sessions-overview"];
    case "session":
      return ["session", route.id];
    case "verify-email":
      return ["emails", route.id, "verify"];
    case "client":
      return ["clients", route.id];
    case "browser-session-list":
      return ["browser-sessions"];
    case "browser-session":
      return ["browser-sessions", route.id];
    case "oauth2-session-list":
      return ["oauth2-sessions"];
    case "compat-session-list":
      return ["compat-sessions"];
    case "unknown":
      return route.segments;
  }
};

const P = Symbol();
type PatternItem = string | typeof P;

// Returns true if the segments match the pattern, where P is a parameter
const segmentMatches = (
  segments: string[],
  ...pattern: PatternItem[]
): boolean => {
  // Quick check to see if the lengths match
  if (segments.length !== pattern.length) return false;

  // Check each segment
  for (let i = 0; i < segments.length; i++) {
    // If the pattern is P, then it's a parameter and we can skip it
    if (pattern[i] === P) continue;
    // Otherwise, check that the segment matches the pattern
    if (segments[i] !== pattern[i]) return false;
  }

  return true;
};

export const segmentsToRoute = (segments: string[]): Route => {
  const matches = (...pattern: PatternItem[]): boolean =>
    segmentMatches(segments, ...pattern);

  // Special case for the home page
  if (segments.length === 0 || (segments.length === 1 && segments[0] === "")) {
    return { type: "profile" };
  }

  if (matches("sessions-overview")) {
    return { type: "sessions-overview" };
  }

  if (matches("browser-sessions")) {
    return { type: "browser-session-list" };
  }

  if (matches("oauth2-sessions")) {
    return { type: "oauth2-session-list" };
  }

  if (matches("compat-sessions")) {
    return { type: "compat-session-list" };
  }

  if (matches("emails", P, "verify")) {
    return { type: "verify-email", id: segments[1] };
  }

  if (matches("clients", P)) {
    return { type: "client", id: segments[1] };
  }

  if (matches("browser-sessions", P)) {
    return { type: "browser-session", id: segments[1] };
  }

  if (matches("session", P)) {
    return { type: "session", id: segments[1] };
  }

  return { type: "unknown", segments };
};

const routeToPath = (route: Route): string =>
  routeToSegments(route)
    .map((part) => encodeURIComponent(part))
    .join("/");

export const appConfigAtom = atom<AppConfig>(
  typeof window !== "undefined" ? window.APP_CONFIG : { root: "/" },
);

const pathToRoute = (path: string): Route => {
  const segments = path.split("/").map(decodeURIComponent);
  return segmentsToRoute(segments);
};

const locationToRoute = (root: string, location: Location): Route => {
  if (!location.pathname || !location.pathname.startsWith(root)) {
    throw new Error(`Invalid location ${location.pathname}`);
  }

  const path = location.pathname.slice(root.length);
  return pathToRoute(path);
};

export const locationAtom = atomWithLocation();
export const routeAtom = atom(
  (get) => {
    const location = get(locationAtom);
    const config = get(appConfigAtom);
    return locationToRoute(config.root, location);
  },
  (get, set, value: Route) => {
    const appConfig = get(appConfigAtom);
    set(locationAtom, {
      pathname: appConfig.root + routeToPath(value),
    });
  },
);

const SessionsOverview = lazy(() => import("./pages/SessionsOverview"));
const SessionDetail = lazy(() => import("./pages/SessionDetail"));
const Profile = lazy(() => import("./pages/Profile"));
const OAuth2Client = lazy(() => import("./pages/OAuth2Client"));
const BrowserSession = lazy(() => import("./pages/BrowserSession"));
const BrowserSessionList = lazy(() => import("./pages/BrowserSessionList"));
const CompatSessionList = lazy(() => import("./pages/CompatSessionList"));
const OAuth2SessionList = lazy(() => import("./pages/OAuth2SessionList"));
const VerifyEmail = lazy(() => import("./pages/VerifyEmail"));

const InnerRouter: React.FC = () => {
  const route = useAtomValue(routeAtom);

  switch (route.type) {
    case "profile":
      return <Profile />;
    case "sessions-overview":
      return <SessionsOverview />;
    case "session":
      return <SessionDetail deviceId={route.id} />;
    case "oauth2-session-list":
      return <OAuth2SessionList />;
    case "browser-session-list":
      return <BrowserSessionList />;
    case "compat-session-list":
      return <CompatSessionList />;
    case "client":
      return <OAuth2Client id={route.id} />;
    case "browser-session":
      return <BrowserSession id={route.id} />;
    case "verify-email":
      return <VerifyEmail id={route.id} />;
    case "unknown":
      return <>Unknown route {JSON.stringify(route.segments)}</>;
  }
};

const Router: React.FC = () => (
  <Layout>
    <Suspense fallback={<LoadingSpinner />}>
      <InnerRouter />
    </Suspense>
  </Layout>
);

// Filter out clicks with modifiers or that have been prevented
const shouldHandleClick = (e: React.MouseEvent): boolean =>
  !e.defaultPrevented &&
  e.button === 0 &&
  !(e.metaKey || e.altKey || e.ctrlKey || e.shiftKey);

export const Link: React.FC<
  {
    route: Route;
    // adds button-like styling to link element
    kind?: "button";
  } & React.HTMLProps<HTMLAnchorElement>
> = ({ route, children, kind, className, ...props }) => {
  const config = useAtomValue(appConfigAtom);
  const path = routeToPath(route);
  const fullUrl = config.root + path;
  const setRoute = useSetAtom(routeAtom);

  // TODO: we should probably have more user control over this
  const [isPending, startTransition] = useTransition();

  const classNames = [
    kind === "button" ? styles.linkButton : "",
    className,
  ].join("");

  return (
    <a
      href={fullUrl}
      onClick={(e: React.MouseEvent): void => {
        // Only handle left clicks without modifiers
        if (!shouldHandleClick(e)) {
          return;
        }

        e.preventDefault();
        startTransition(() => {
          setRoute(route);
        });
      }}
      className={classNames}
      {...props}
    >
      {isPending ? "Loading..." : children}
    </a>
  );
};

export default Router;
