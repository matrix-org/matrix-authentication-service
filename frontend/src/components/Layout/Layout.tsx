// Copyright 2022 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { useAtomValue } from "jotai";
import { useTranslation } from "react-i18next";

import { currentUserIdAtom } from "../../atoms";
import { isErr, unwrapErr, unwrapOk } from "../../result";
import { appConfigAtom, routeAtom } from "../../routing";
import Footer from "../Footer";
import GraphQLError from "../GraphQLError";
import NavBar from "../NavBar";
import NavItem from "../NavItem";
import NotLoggedIn from "../NotLoggedIn";
import UserGreeting from "../UserGreeting";

import styles from "./Layout.module.css";

const Layout: React.FC<{ children?: React.ReactNode }> = ({ children }) => {
  const route = useAtomValue(routeAtom);
  const appConfig = useAtomValue(appConfigAtom);
  const result = useAtomValue(currentUserIdAtom);
  const { t } = useTranslation();

  if (isErr(result)) return <GraphQLError error={unwrapErr(result)} />;

  // Hide the nav bar & user greeting on the verify-email page
  const shouldHideNavBar = route.type === "verify-email";

  const userId = unwrapOk(result);
  if (userId === null)
    return (
      <div className={styles.container}>
        <NotLoggedIn />
      </div>
    );

  return (
    <div className={styles.layoutContainer}>
      {shouldHideNavBar ? null : (
        <>
          <UserGreeting userId={userId} />

          <NavBar>
            <NavItem route={{ type: "profile" }}>
              {t("frontend.nav.profile")}
            </NavItem>
            <NavItem route={{ type: "sessions-overview" }}>
              {t("frontend.nav.sessions")}
            </NavItem>
          </NavBar>
        </>
      )}

      <main>{children}</main>

      <Footer
        imprint={appConfig.branding?.imprint}
        tosUri={appConfig.branding?.tosUri}
        policyUri={appConfig.branding?.policyUri}
      />
    </div>
  );
};

export default Layout;
