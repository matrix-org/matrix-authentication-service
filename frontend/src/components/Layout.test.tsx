// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// @vitest-environment happy-dom

import { render } from "@testing-library/react";
import { Provider } from "jotai";
import { useHydrateAtoms } from "jotai/utils";
import { Suspense } from "react";
import { describe, expect, it, vi, afterAll, beforeEach } from "vitest";

import { appConfigAtom, locationAtom } from "../Router";
import { currentUserIdAtom, GqlResult } from "../atoms";

import Layout from "./Layout";

beforeEach(async () => {
  // For some reason, the locationAtom gets updated with `about:black` on render,
  // so we need to set a "real" location and wait for the next tick
  window.location.assign("https://example.com/");
  // Wait the next tick for the location to update
  await new Promise((resolve) => setTimeout(resolve, 0));
});

const HydrateLocation: React.FC<React.PropsWithChildren<{ path: string }>> = ({
  children,
  path,
}) => {
  useHydrateAtoms([
    [appConfigAtom, { root: "/" }],
    [locationAtom, { pathname: path }],
  ]);
  return <>{children}</>;
};

const WithLocation: React.FC<React.PropsWithChildren<{ path: string }>> = ({
  children,
  path,
}) => {
  return (
    <Provider>
      <Suspense>
        <HydrateLocation path={path}>{children}</HydrateLocation>
      </Suspense>
    </Provider>
  );
};

describe("<Layout />", () => {
  beforeEach(() => {
    vi.spyOn(currentUserIdAtom, "read").mockResolvedValue(
      "abc123" as unknown as GqlResult<string | null>,
    );
  });
  afterAll(() => {
    vi.restoreAllMocks();
  });
  it("renders app navigation correctly", async () => {
    const component = render(
      <WithLocation path="/account">
        <Layout />
      </WithLocation>,
    );

    expect(await component.findByText("Profile")).toMatchSnapshot();
    expect(await component.findByText("Home")).toMatchSnapshot();
  });
});
