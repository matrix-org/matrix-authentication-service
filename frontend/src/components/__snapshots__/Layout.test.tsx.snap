// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Layout /> > renders app navigation correctly 1`] = `
<a
  class="_navItem_8603fc"
  href="/profile"
>
  Profile
</a>
`;

exports[`<Layout /> > renders app navigation correctly 2`] = `
<a
  aria-current="page"
  class="_navItem_8603fc"
  href="/"
>
  Home
</a>
`;
